src/clients.rs

```diff
@@ -125,20 +128,123 @@ impl From<tonic::Status> for Error {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub enum ClientCode {
+    Http(StatusCode),
+    Grpc(tonic::Code),
+}
+
+impl Display for ClientCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ClientCode::Http(code) => write!(f, "HTTP {}", code),
+            ClientCode::Grpc(code) => write!(f, "gRPC {:?} {}", code, code),
+        }
+    }
+}
+
 #[derive(Clone)]
 pub struct HttpClient {
     base_url: Url,
+    health_url: Url,
     client: reqwest::Client,
 }
 
 impl HttpClient {
     pub fn new(base_url: Url, client: reqwest::Client) -> Self {
-        Self { base_url, client }
+        let health_url = extract_base_url(&base_url).join("health").unwrap();
+        Self {
+            base_url,
+            health_url,
+            client,
+        }
     }
 
     pub fn base_url(&self) -> &Url {
         &self.base_url
     }
+
+    /// This is sectioned off to allow for testing.
+    pub(super) async fn http_response_to_health_check_result(
+        res: Result<Response, reqwest::Error>,
+    ) -> HealthCheckResult {
```

Comment:
All of the logic in this method feels unnecessary IMO. I think we can simply consider a non-200 status code on a health endpoint to be unhealthy? 

Also, I think `HealthStatus::NotReady` should be renamed to `HealthStatus::Unhealthy`, as we are checking health, not readiness, and `HealthStatus::Unknown` should be dropped as a service is either healthy or it isn't, why introduce the concept of "unknown" here? Unknown essentially means unhealthy in this context.